### PR TITLE
Two-factor authenticaton  closes #1286 

### LINKED ISSUE #1286: Add optional 2FA
This would make sense to be added in the `certego_saas` as a separate application.

### PR DESCRIPTION
**Pull request by shivam-Purohit**
Two-factor authentication

# Description

Optional 2fa using the Django-two-factor auth package.

## Type of change

- [x] New feature (non-breaking change which adds functionality).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [ ] The pull request is for the branch `develop`
- [ ] A new plugin (analyzer, connector, visualizer or playbook) was added or changed, in which case:
    - [ ] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](./Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [ ] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [ ] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [ ] If you created a new analyzer and it is free (does not require API keys), please add it in the `FREE_TO_USE_ANALYZERS` playbook in `playbook_config.json`.
    - [ ] Check if it could make sense to add that analyzer/connector to other freely available playbooks.
    - [ ] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
- [ ] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](./Contribute.md)).
- [x] If the GUI has been modified:
    - [x] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.

### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.

### CODE REVIEW

**Comment by shivam-Purohit** [comment]
I don't know if the whole app is required to add but if we need to change the templates, related views also need to be changed.

File: docker/env_file_app_ci [hunk 1]
```diff
@@ -56,5 +56,5 @@ BROKER_URL=amqp://guest:guest@rabbitmq:5672
 
 
 # drf-recaptcha
-RECAPTCHA_SECRET_KEY_IO_PUBLIC="testkey"
-RECAPTCHA_SECRET_KEY_IO_LOCAL="testkey"
+ RECAPTCHA_SECRET_KEY_IO_PUBLIC="testkey"
```
Comment (mlodic) [inline docker/env_file_app_ci:59]: I guess this space is a leftover from a comment, could you remove it?

**Comment by mlodic** [comment]
thank you for having shared your first PR for the GSoC :) we'll review it soon.
Just one thing to remember that I asked you: please create branches from the original IntelOwl repo and not your fork, otherwise it is complicated from us to work on your Pull Request and help you. If you can, please close this PR and open a new one from IntelOwl itself. thank you

**Review by mlodic — COMMENTED** [review]
why did you copy/paste the entire `two_factor` package from the official library here?

**Comment by shivam-Purohit** [comment]
> why did you copy/paste the entire `two_factor` package from the official library here?

To change the templates we need to copy/paste the original one. But we know we need to change those to react components. So we need to make some changes to the views as well. So I copy paste the whole app. 
I am bit confused which one to keep and which to not

File: docker/env_file_app_ci (hunk 1 above)
Comment (shivam-Purohit) [reply docker/env_file_app_ci:59]: yes, I made unrelated changes. I will cover them at once.

**Comment by shivam-Purohit** [comment]
@mlodic I tried that but I guess it's the only way you can customize the app's frontend. I could not find information to do that, the only one I found was copying the template folder and changing the templates.
But as we need the templates to be converted to react components the corresponding views need to render react.  Some logic also needs to be rewritten in views for all the files in the folder.

**Comment by mlodic** [comment]
I see that the django-two-factor-auth library is not meant to be used in a project that leverages a frontend which is not the standard one for Django. 

I guess this leaves us to these possible chances:
* you try to do something with `django-otp` lower level library and see if you can make it work
* you skip this issue for now and you go forward with the other ones which are already planned for the GSoC and get back to this one later. In this way you are not stuck with an issue that is more difficult than expected.
* we select other issues as a substitute if this is still too difficult. We have a lot of analyzers with are yet to be implemented and that would help us to remove some issues from the backlog.

Please let me know

**Comment by shivam-Purohit** [comment]
I can try out something with django-otp but I will need some time. I will move forward with my project and try out django-otp in the meantime. As some of the issues that were listed in the proposal are already solved or removed I can surely work on some substitute issues once I am done with the original ones.

**Comment by shivam-Purohit** [comment]
@mlodic should I close this PR?

**Pull request closed without merging**